feat(native restaking): add a function to update the states of native token restaking

TimmyExogenous opened the pull request:
## Description

This PR adds a function to update the state when Oracle updates the balance of the native restaking token.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new method for iterating over undelegation records by staker and asset.
	- Added functionality to manage native restaking balances for stakers and assets.

- **Improvements**
	- Enhanced error handling across various functions to provide clearer messages and control flow.
	- Updated delegation operations to allow for more nuanced success and error reporting.
	- Renamed functions for improved clarity and readability.
	- Streamlined test preparation processes for better parameterization.

- **Bug Fixes**
	- Modified logic in share management to correctly reflect changes in operator shares.

- **Documentation**
	- Improved clarity of error messages for better debugging and user experience.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

TimmyExogenous committed:
detail the return error for operator info

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
add a function to update the states of native token restaking

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

leonz789 commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -12,7 +12,7 @@ import (
+type DelegationOpFunc func(keys *delegationtype.SingleDelegationInfoReq, amounts *delegationtype.DelegationAmounts) (bool, error)
 
 // IterateDelegationsForStakerAndAsset processes all operations
 // that require iterating over delegations for a specified staker and asset.
```
Can we refactor `IterateDelegationsForStakerAndAsset` and `IterateDelegationsForStaker` into one method ?

TimmyExogenous replied:
Thanks, I will keep these two interfaces and use a common function to refactor them.

leonz789 commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -60,22 +66,22 @@ func (k Keeper) IterateDelegationsForStaker(ctx sdk.Context, stakerID string, op
+		return false, nil
 	}
 	err = k.IterateDelegationsForStakerAndAsset(ctx, stakerID, assetID, opFunc)
 	if err != nil {
```
these 4 lines equals to 
`return amount, err`

TimmyExogenous replied:
Thanks, I will refine it in the next commit.

leonz789 commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -60,22 +66,22 @@ func (k Keeper) IterateDelegationsForStaker(ctx sdk.Context, stakerID string, op
+	opFunc := func(keys *delegationtype.SingleDelegationInfoReq, amounts *delegationtype.DelegationAmounts) (bool, error) {
 		if amounts.UndelegatableShare.IsZero() {
-			return nil
+			return true, nil
```
`return false, nil`
iteration should move on to next operator

TimmyExogenous replied:
Thanks, I will fix it in the next commit.

leonz789 commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -88,19 +94,19 @@ func (k Keeper) StakerDelegatedTotalAmount(ctx sdk.Context, stakerID string, ass
 // the key of return value is the operator address, and the value is the asset amount.
 func (k *Keeper) AllDelegatedAmountForStakerAsset(ctx sdk.Context, stakerID string, assetID string) (map[string]sdkmath.Int, error) {
```
The name is kind of confused with `StakerDelegatedTotalAmount`.
Maybe changed to some name like `DelegatedAmountPerOperatorForStakerAsset`.

TimmyExogenous replied:
Thanks, I will rename them to `TotalDelegatedAmountForStakerAsset` and `AllDelegatedInfoForStakerAsset` in the next commit.

leonz789 commented on x/delegation/keeper/delegation_state.go:
```diff
@@ -177,9 +183,9 @@ func (k *Keeper) GetSingleDelegationInfo(ctx sdk.Context, stakerID, assetID, ope
 func (k *Keeper) GetDelegationInfo(ctx sdk.Context, stakerID, assetID string) (*delegationtype.QueryDelegationInfoResponse, error) {
 	var ret delegationtype.QueryDelegationInfoResponse
 	ret.DelegationInfos = make(map[string]*delegationtype.DelegationAmounts, 0)
```
equals to `make(string]*delegationtype.DelegationAmounts)`

TimmyExogenous replied:
right, I will change it in the next commit.

leonz789 commented on x/delegation/keeper/un_delegation_state.go:
```diff
@@ -150,6 +150,37 @@ func (k *Keeper) GetStakerUndelegationRecords(ctx sdk.Context, stakerID, assetID
+	opFunc func(undelegation *types.UndelegationRecord) (bool, error),
+) error {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixStakerUndelegationInfo)
+	iterator := sdk.KVStorePrefixIterator(store, []byte(strings.Join([]string{stakerID, assetID}, "/")))
```
use `delegationtype.GetDelegationStateIteratorPrefix(stakerID, assetID)` ?

TimmyExogenous replied:
Right, I will rename the function `GetDelegationStateIteratorPrefix` and use it here.

MaxMustermann2 commented on x/delegation/keeper/update_native_restaking_balance.go:
```diff
@@ -0,0 +1,112 @@
+		// and the delegated share last if there is still a remaining amount that needs to be slashed.
+		// When slash from the delegated share, we can slash sequentially from all delegated operators
+		// according to the order of the operator addresses in the KV store. We can also proportionally
+		// slash from all operators to whom the related staker has already delegated, if we think
+		// this approach is better
```
I think this approach is better.

Say that native staked asset deposited by staker S is 64 ETH of which 32 ETH is free, 16 ETH is delegated to operator A and 16 ETH is unbonding from operator B. In this scenario, staker S is running a validator on Ethereum but may not necessarily be running one on Exocore or any of its AVSs.

The native staked ETH is slashed in half to 32 ETH. This half should be applied to each of the items:
- Free deposit -> 32 ETH becomes 16 ETH
- Delegated to A -> 16 ETH becomes 8 ETH and the vote power change is applied immediately across the AVSs in which A has opted.
- Unbonding from B -> 16 ETH becomes 8 ETH and continues unbonding according to the same schedule

TimmyExogenous replied:
I will adjust the implementation when slashing from the delegated amount.

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
address the issues identified by code comments

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
fix the issue identified by code comments

TimmyExogenous committed:
add test for UpdateNativeRestakingBalance

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
fix the issues identified by code comments
adjust the implementation of UpdateNativeRestakingBalance and add the unit test for it

TimmyExogenous commented on x/delegation/keeper/update_native_restaking_balance_test.go:
```diff
@@ -0,0 +1,79 @@
+	anotherOperatorAsset, err := suite.App.AssetsKeeper.GetOperatorSpecifiedAssetInfo(suite.Ctx, anotherOperatorAddr, assetID)
+	suite.NoError(err)
+	suite.Equal(sdkmath.NewInt(14), anotherOperatorAsset.TotalAmount)
+	suite.Equal(expectedShareForAnotherOpeartor, anotherOperatorAsset.TotalShare)
```
Seems like it's unnecessary to refactor.

TimmyExogenous commented on x/delegation/keeper/update_native_restaking_balance_test.go:
```diff
@@ -0,0 +1,79 @@
+	delegationForDefaultOperator, err := suite.App.DelegationKeeper.GetSingleDelegationInfo(suite.Ctx, stakerID, assetID, suite.opAccAddr.String())
+	suite.NoError(err)
+	slashProportion := sdkmath.LegacyNewDec(1).Sub((sdkmath.LegacyNewDec(1)).Quo(sdkmath.LegacyNewDec(3)))
+	expectedShareForDefaultOpeartor := sdkmath.LegacyNewDec(10).Mul(slashProportion)
```
thanks

TimmyExogenous commented on x/delegation/keeper/update_native_restaking_balance_test.go:
```diff
@@ -0,0 +1,79 @@
+
+	delegationForAnotherOperator, err := suite.App.DelegationKeeper.GetSingleDelegationInfo(suite.Ctx, stakerID, assetID, anotherOperatorAddr.String())
+	suite.NoError(err)
+	expectedShareForAnotherOpeartor := sdkmath.LegacyNewDec(20).Mul(slashProportion)
```
thanks

TimmyExogenous commented on x/delegation/keeper/update_native_restaking_balance.go:
```diff
@@ -0,0 +1,129 @@
+		// whom the related staker has already delegated
+		if remainAmount.IsPositive() {
+			// calculate the slash proportion
+			totalDelegatedAmount, err := k.TotalDelegatedAmountForStakerAsset(ctx, stakerID, assetID)
+			if err != nil {
+				return err
+			}
+			slashProportion := sdkmath.LegacyNewDecFromBigInt(remainAmount.BigInt()).Quo(sdkmath.LegacyNewDecFromBigInt(totalDelegatedAmount.BigInt()))
+			if slashProportion.GT(sdkmath.LegacyNewDec(1)) {
+				slashProportion = sdkmath.LegacyNewDec(1)
+			}
```
Thanks, I will address this issue in the next commit.

TimmyExogenous commented on x/operator/keeper/usd_value.go:
```diff
@@ -469,30 +469,30 @@ func (k *Keeper) CalculateUSDValueForStaker(ctx sdk.Context, stakerID, avsAddr s
 				price, ok := prices[keys.AssetID]
 				if !ok {
-					return errorsmod.Wrapf(operatortypes.ErrKeyNotExistInMap, "CalculateUSDValueForStaker assetID doesn't exist, assetID:%s", keys.AssetID)
+					return true, errorsmod.Wrapf(operatortypes.ErrKeyNotExistInMap, "CalculateUSDValueForStaker assetID doesn't exist, assetID:%s", keys.AssetID)
```
Thanks, I will update it in the next commit.

TimmyExogenous commented on x/operator/keeper/usd_value.go:
```diff
@@ -469,30 +469,30 @@
 			if _, ok := assets[keys.AssetID]; ok {
 				price, ok := prices[keys.AssetID]
 				if !ok {
-					return errorsmod.Wrapf(operatortypes.ErrKeyNotExistInMap, "CalculateUSDValueForStaker assetID doesn't exist, assetID:%s", keys.AssetID)
+					return true, errorsmod.Wrapf(operatortypes.ErrKeyNotExistInMap, "CalculateUSDValueForStaker assetID doesn't exist, assetID:%s", keys.AssetID)
 				}
 				operatorAsset, err := k.assetsKeeper.GetOperatorSpecifiedAssetInfo(ctx, operator, keys.AssetID)
 				if err != nil {
-					return err
+					return true, err
 				}
 				amount, err := delegationkeeper.TokensFromShares(amounts.UndelegatableShare, operatorAsset.TotalShare, operatorAsset.TotalAmount)
 				if err != nil {
-					return err
+					return true, err
 				}
 				assetInfo, err := k.assetsKeeper.GetStakingAssetInfo(ctx, keys.AssetID)
 				if err != nil {
-					return err
+					return true, err
 				}
 				usdValue := CalculateUSDValue(amount, price.Value, assetInfo.AssetBasicInfo.Decimals, price.Decimal)
 				totalUSDValue = totalUSDValue.Add(usdValue)
 			}
 		}
-		return nil
+		return false, nil
```
Thanks, I will add the code comment in the next commit.

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

TimmyExogenous committed:
rebase to develop branch and fix the conflict

adu-web3 reviewed (approved):
LGTM

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop

bwhour commented on x/delegation/keeper/update_native_restaking_balance.go:
```diff
@@ -0,0 +1,131 @@
+			}
+			if !totalDelegatedAmount.IsZero() {
+				slashProportion := sdkmath.LegacyNewDecFromBigInt(remainAmount.BigInt()).Quo(sdkmath.LegacyNewDecFromBigInt(totalDelegatedAmount.BigInt()))
+				if slashProportion.GT(sdkmath.LegacyNewDec(1)) {
```
Replace this magic numbers with constants

> const (
>     MaxSlashProportion = 1
> )

TimmyExogenous replied:
ok

bwhour commented on x/delegation/keeper/update_native_restaking_balance.go:
```diff
@@ -0,0 +1,131 @@
+						return true, err
+					}
+					slashShare := delegationAmount.UndelegatableShare.Mul(slashProportion)
+					actualSlashAmount, err := k.RemoveShare(ctx, false, opAccAddr, stakerID, assetID, slashShare)
```
can we extract reusable logic into standalone functions ? like 

> func (k Keeper) slashFromWithdrawable(ctx sdk.Context, stakerID, assetID string, amount sdkmath.Int) (sdkmath.Int, error) {
    // Implement the logic to deduct from withdrawable amount
}

>func (k Keeper) slashFromUndelegations(ctx sdk.Context, stakerID, assetID string, amount sdkmath.Int) (sdkmath.Int, error) {
    // Implement the logic to deduct from the pending undelegated amount.
}

>func (k Keeper) slashFromDelegatedShare(ctx sdk.Context, stakerID, assetID string, amount sdkmath.Int) (sdkmath.Int, error) {
    // Implement the logic to deduct from the delegated share.
}

TimmyExogenous replied:
I will record this issue and change it in the future.

bwhour commented on x/delegation/keeper/update_native_restaking_balance.go:
```diff
@@ -0,0 +1,131 @@
+					if err != nil {
+						return true, err
+					}
+					remainAmount = remainAmount.Sub(actualSlashAmount)
```
it is better to use _pendingSlashAmount_ to replace remainAmount

TimmyExogenous replied:
fixed

bwhour reviewed (approved):
Add some comment. the overall looks good.

TimmyExogenous committed:
add a const MaxSlashProportion and rename the remainingSlashAmount to pendingSlashAmount

TimmyExogenous committed:
Merge remote-tracking branch 'upstream/develop' into develop
